### ISSUE TITLE
Connection was too slow.(ConnectionType.MANUAL)

### ISSUE THREAD

**Issue by devshy123** [WalletConnect/WalletConnectKotlinV2#687]
I'm linking with the bitkeep app. (MANUAL method)

CoreClient.initialize(
relayServerUrl = serverUrl,
connectionType = ConnectionType.MANUAL,
application = this,
metaData = appMetaData
) { error ->
}

The connection time has suddenly become too slow since yesterday.
It takes more than 10 to 15 seconds.
Is there any solution?

https://user-images.githubusercontent.com/124105699/221363629-acbb4902-a723-4de9-9926-703727629f27.mp4

**Comment by TalhaAli00** [comment]
@devshy123 
Thanks for reporting this. I have a few questions

- where are you located
- what version of the SDKs are you using
- do you notice a change when using `AUTOMATIC` mode

**Comment by arein** [comment]
Do you have any identifiers e.g. RPC id or project Id you can share? Feel free to dm

**Comment by arein** [comment]
I looked up the logs for BitKeep and it looks like you're using Core SDK version `1.3.0`. Unless you have a different testing project id.

Please upgrade to the latest version and report again https://github.com/WalletConnect/WalletConnectKotlinV2
Should be version `1.11.0`.

**Comment by devshy123** [comment]
@TalhaAli00
@arein


where are you located
 - South Korea 

what version of the SDKs are you using
 - I am using the 1.4.1 version.
 - I know that the 1.6.0. version was released, but the event didn't work when I used the higher version than 1.4.1. on the Bitkeep app.
    override fun onSessionApproved(approvedSession: Sign.Model.ApprovedSession) 

do you notice a change when using AUTOMATIC mode
- This issue does not occur when using automatic but another issue occurred.
  Another issue: Run approval button -> No response -> Normal operation when you return to my app
- For more information, refer to the video linked using automatic.

https://user-images.githubusercontent.com/124105699/221384683-9f1a3450-b50c-4aa3-836b-9e2a665dc757.mp4

**Comment by arein** [comment]
You need to use the latest version. What issues are you seeing?

**Comment by devshy123** [comment]
> * I know that the 1.6.0. version was released, but the event didn't work when I used the higher version than 1.4.1. on the Bitkeep app.
>   override fun onSessionApproved(approvedSession: Sign.Model.ApprovedSession)

**Comment by china-mobile2008** [comment]
> @devshy123 Thanks for reporting this. I have a few questions
> 
> * where are you located
> * what version of the SDKs are you using
> * do you notice a change when using `AUTOMATIC` mode
> 
> CC @arein

same issue,  I found that the latest official BOM 1.6.0 signature and authorization cannot connect BitKeep wallet, but the signature of BOM 1.4.1 can connect BitKeep, while the authorization cannot and it remains in a waiting state before prompting a connection failure.

where are you located

- china


https://user-images.githubusercontent.com/10908640/221389357-9a86742b-263a-493d-8d42-c27758c3fa08.mp4

**Comment by arein** [comment]
What endpoint are you using? Are you behind VPN? Non VPN in China needs to use `walletconnect.org` not `walletconnect.com`.

Otherwise can you please provide more debug logs to enable us to provider better help.

**Comment by china-mobile2008** [comment]
> walletconnect.org
@arein 

I have been using walletconnect.org and it takes a lot of time and has a very low success rate every time I sign. I would like to ask how much time it takes for each signature authorization on your end. Even when using a VPN, the results are the same. Could you please confirm what modifications have been made in BOM 1.4.1 and BOM 1.6.0? With the 1.6.0 version, both signature and authorization cannot connect to the BitKeep wallet.

**Comment by arein** [comment]
@china-mobile2008 can you share any identifiers so we can look on our end?
You need to be more specific in your bug reports please to help us investigate.

> it takes a lot of time and has a very low success rate every time I sign
Please quantify both of these statements - how long and how low?

> I would like to ask how much time it takes for each signature authorization on your end
the entire flow of connecting a wallet and signing a message takes 2s as per our public metrics: https://status.walletconnect.com/

> Could you please confirm what modifications have been made in BOM 1.4.1 and BOM 1.6.0
https://github.com/WalletConnect/WalletConnectKotlinV2/releases

> both signature and authorization cannot connect to the BitKeep wallet
You have to provide better error reports and logs here. This is not an actionable statement for us. We want to help you. Please enable us to do so by putting in the effort of providing a good error report.
What is the exact expected vs. actual behavior and what log statements do you have to support it.

**Comment by china-mobile2008** [comment]
> @china-mobile2008 can you share any identifiers so we can look on our end? You need to be more specific in your bug reports please to help us investigate.
> 
> > it takes a lot of time and has a very low success rate every time I sign
> 
> Please quantify both of these statements - how long and how low?
> 
> > I would like to ask how much time it takes for each signature authorization on your end
> 
> the entire flow of connecting a wallet and signing a message takes 2s as per our public metrics: https://status.walletconnect.com/
> 
> > Could you please confirm what modifications have been made in BOM 1.4.1 and BOM 1.6.0
> 
> https://github.com/WalletConnect/WalletConnectKotlinV2/releases
> 
> > both signature and authorization cannot connect to the BitKeep wallet
> 
> You have to provide better error reports and logs here. This is not an actionable statement for us. We want to help you. Please enable us to do so by putting in the effort of providing a good error report. What is the exact expected vs. actual behavior and what log statements do you have to support it.

@arein 
I provided a log and hope it's helpful to you. Could you provide a demo for the manual mode? I found that changing to manual mode improves the success rate when connecting to the wallet. Also, I have a question: does the Dapp connect to Bitkeep through relay services to relay some messages? Is this the reason for slow access? Can we customize the local WebSocket? If possible, please provide a demo. Thank you.

[log.txt](https://github.com/WalletConnect/WalletConnectKotlinV2/files/10881570/log.txt)

**Comment by jakubuid** [comment]
Hi @china-mobile2008 let me have a look on that

**Comment by china-mobile2008** [comment]
@jakubuid 

**The problem is that when I click on the authorization button, it stays in a waiting state and I don't know whether it's an issue with WalletConnect or BitKeep.**

```
2023-03-07 12:46:24.858 11249-11249 HwResolver              system:ui                            D  ResolveListAdapter: initCurrentResolveList currentResolveList is [ResolveInfo{6c72e3d com.bitkeep.w5/.MainActivity m=0x208000}, ResolveInfo{3137632 com.wallet.crypto.trustapp/.ui.start.activity.RootHostActivity m=0x208000}, ResolveInfo{9f8bf83 com.walletconnect.responder/.ResponderActivity m=0x208000}, ResolveInfo{b95a500 com.walletconnect.wallet/.ui.host.WalletSampleActivity m=0x208000}, ResolveInfo{4fd7139 com.walletconnect.web3.wallet.sample/com.walletconnect.web3.wallet.ui.Web3WalletActivity m=0x208000}, ResolveInfo{37ec37e im.token.app/org.consenlabs.imtoken.MainActivity m=0x208000}, ResolveInfo{dfdd0df io.horizontalsystems.bankwallet/.MainLauncherAlias m=0x208000}, ResolveInfo{630112c io.horizontalsystems.bankwallet.dev/.MainLauncherAlias m=0x208000}, ResolveInfo{bbae7f5 org.bitizen.wallet/.MainActivity m=0x208000}, ResolveInfo{dbf998a so.onekey.app.wallet/.MainActivity m=0x208000}, ResolveInfo{6189bfb vip.mytokenpocket/com.tokenbank.pull.PullUpActivity m=0x208000}]
2023-03-07 12:46:24.858 11249-11249 HwResolver              system:ui                            W  ResolveListAdapter:ShareList need show: [com.bitkeep.w5.MainActivity, com.wallet.crypto.trustapp.ui.start.activity.RootHostActivity, com.walletconnect.responder.ResponderActivity, com.walletconnect.wallet.ui.host.WalletSampleActivity, com.walletconnect.web3.wallet.ui.Web3WalletActivity, org.consenlabs.imtoken.MainActivity, io.horizontalsystems.bankwallet.MainLauncherAlias, io.horizontalsystems.bankwallet.dev.MainLauncherAlias, org.bitizen.wallet.MainActivity, so.onekey.app.wallet.MainActivity, com.tokenbank.pull.PullUpActivity]
2023-03-07 12:46:24.945 11249-28091 HwResourcesImpl         system:ui                            W  icon in pkg com.walletconnect.responder is too large.sDefaultSizeWithoutEdge = 189 bmp.getWidth() = 1536
2023-03-07 12:46:24.983 11249-27569 HwResourcesImpl         system:ui                            W  icon in pkg com.walletconnect.wallet is too large.sDefaultSizeWithoutEdge = 189 bmp.getWidth() = 1536
2023-03-07 12:46:28.048 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:28.049 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: DefaultDispatcher-worker-3 [onConnectionStateChange(WalletDelegate.kt:100)]
2023-03-07 12:46:28.049 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ║ onConnectionStateChange(ConnectionState(isAvailable=true))
2023-03-07 12:46:28.049 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:28.049 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:28.050 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: DefaultDispatcher-worker-3 [runMainThread(WalletConnectV2Plugin.kt:87)]
2023-03-07 12:46:28.050 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ║ runMainThread:map = {onConnected=true},method =onConnected
2023-03-07 12:46:28.050 28098-28154 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:29.614 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:29.622 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: DefaultDispatcher-worker-21 [onSessionProposal(WalletDelegate.kt:28)]
2023-03-07 12:46:29.622 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ║ onSessionProposal
2023-03-07 12:46:29.622 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:29.668 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:29.668 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: DefaultDispatcher-worker-21 [runMainThread(WalletConnectV2Plugin.kt:87)]
2023-03-07 12:46:29.668 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ║ runMainThread:map = {topic=, proposer={"description":"MarsBit Dapp","icons":[""],"name":"MarsBit","proposerPublicKey":"c0dbccafc3abbd2e2c5889f8043080c4dc8e03e00dc3d4719ae30a51426ce667","relayProtocol":"irn","url":"marsbit.io"}, requirednamespaces={"eip155":{"chains":["eip155:137"],"events":["chainChanged","accountChanged"],"methods":["eth_sendTransaction","personal_sign","eth_sign","eth_signTypedData"]}}},method =onProposal
2023-03-07 12:46:29.669 28098-28472 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:30.763 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: main [approve(WalletConnectV2Plugin.kt:147)]
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ║ sessionProposal =SessionProposal(name=MarsBit, description=MarsBit Dapp, url=marsbit.io, icons=[], requiredNamespaces={eip155=Proposal(chains=[eip155:137], methods=[eth_sendTransaction, personal_sign, eth_sign, eth_signTypedData], events=[chainChanged, accountChanged], extensions=null)}, proposerPublicKey=c0dbccafc3abbd2e2c5889f8043080c4dc8e03e00dc3d4719ae30a51426ce667, relayProtocol=irn, relayData=null)
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ╔═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ║ In Thread: main [approve(WalletConnectV2Plugin.kt:177)]
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ║ approve ={eip155=Session(accounts=[eip155:137:0x28B67f5A03F4476417712E47a0Aa6863a543bd5D], methods=[eth_sendTransaction, personal_sign, eth_sign, eth_signTypedData], events=[chainChanged, accountChanged], extensions=null)}
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  ╚═══════════════════════════════════════════════════════════════════════════════════════════════════
2023-03-07 12:46:30.764 28098-28098 WalletConnectV2         com.bitkeep.w5                       D  approve success ={eip155=Session(accounts=[eip155:137:0x28B67f5A03F4476417712E47a0Aa6863a543bd5D], methods=[eth_sendTransaction, personal_sign, eth_sign, eth_signTypedData], events=[chainChanged, accountChanged], extensions=null)}.toString()

```

**Comment by jakubuid** [comment]
> I provided a log and hope it's helpful to you. Could you provide a demo for the manual mode? I found that changing to manual mode improves the success rate when connecting to the wallet. Also, I have a question: does the Dapp connect to Bitkeep through relay services to relay some messages? Is this the reason for slow access? Can we customize the local WebSocket? If possible, please provide a demo. Thank you.

There should be no difference regarding using the manual or automatic connection mode, the success rate should be the same there is no correlation. It gets to be something different.
Yes, Dapp connects a wallet through the wss connection to relay messages on particular subscriptions.
There is a way to do a custom wss connection by implementing the RelayConnectionInterface and injecting it into the CoreClient initialization method.

Before doing so, I recommend doing more research on your current issues. Hence, is your project open source so I can look at your setup?
Or I'm happy to schedule a pair programming session to see what's on your end.

@china-mobile2008 let know what's the best for you please

**Comment by jakubuid** [comment]
Also, from the logs, I can see that you're using one of the oldest versions of our SDK. 1.4.1 is the last version that does not support [CAIP-25 ](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-25.md) where I would find the issues because the validation and the namespace structure have been changed.

Can you please share the structure of the session proposal and the approved parameters that you're sending after updating to the most up-to-date SDK version? -> BOM 1.6.1

**Comment by jakubuid** [comment]
Hi @china-mobile2008 any updates? I'd love to help and make your use case rock solid :)

**Comment by china-mobile2008** [comment]
hey @jakubuid 

I'm sorry for the delay in response.

 I used the official website's dapp   [here](https://github.com/WalletConnect/WalletConnectKotlinV2/tree/develop/sign/dapp) to link BitKeep. After multiple tests, I found that the following steps can reproduce this issue:

**1.  starting from the dapp, connect to BitKeep manually cancel the return during the connection process and repeat this 
        step multiple times. 
2. Try to connect to the BitKeep wallet again and wait for the connection to succeed. 
4. Disconnect the wallet connection again. 
5. Connect to the BitKeep wallet again.**

and you will find that the BitKeep wallet is in a long waiting state, even if you jump to the authorization page, the authorization button will be in a waiting state for a long time after you click it. The solution to this problem is to clear all data of the BitKeep App. After doing this, you will find that the connection to BitKeep will be very fast again and the success rate will be 100%.

I think the reason why connecting to BitKeep using relay.walletconnect.org is slow is due to BitKeep itself.

**Comment by jakubuid** [comment]
Hey @china-mobile2008 

I was able to reproduce the issue that you're facing.

Firstly, you should update BitKeep wallet to use the most up-to-date SDK version which is  BOM 1.6.1. See the version chart: https://github.com/WalletConnect/WalletConnectKotlinV2#sdk-chart.

- The reason why you're not getting the success callback in the `fun approveSession(approve: Sign.Params.Approve, onSuccess: (Sign.Params.Approve) -> Unit, onError: (Sign.Model.Error) -> Unit)` is that the namespaces in the Approve params have a wrong structure. Please add the proper error handling, because the onError callback should return an error in this case.

Kotlin Dapp sends in session proposal:
```
{
  "requiredNamespaces": {
    "eip155": {
      "chains": [
        "eip155:1"
      ],
      "methods": [
        "eth_sendTransaction",
        "personal_sign",
        "eth_sign",
        "eth_signTypedData"
      ],
      "events": [
        "chainChanged",
        "accountChanged"
      ]
    }
  },
  "optionalNamespaces": {}
}
```
where BitKeep wallet responds:
```
{
  "namespaces": {
    "eip155": {
      "accounts": [
        "eip155:1:0x053F44c2c99A33ef861D21Ae0ccBa93b6021dEeC"
      ],
      "methods": [
        "eth_sendTransaction",
        "personal_sign",
        "eth_sign",
        "eth_signTypedData"
      ],
      "events": [
        "chainChanged",
        "accountChanged"
      ]
    }
  }
}
```
which is incorrect according to the CAIP-25 standard. It should include the chains list. Here you can find more info about it: https://gist.github.com/jakubuid/a5f3fdf61f6a46c6e2bb86468cd4ccbf

The correct namespaces structure in the session approval should be:

```
{
  "namespaces": {
    "eip155": {
      "chains": [
        "eip155:1"
      ],
      "accounts": [
        "eip155:1:0x053F44c2c99A33ef861D21Ae0ccBa93b6021dEeC"
      ],
      "methods": [
        "eth_sendTransaction",
        "personal_sign",
        "eth_sign",
        "eth_signTypedData"
      ],
      "events": [
        "chainChanged",
        "accountChanged"
      ]
    }
  }
}
```
Please add this fix and improve error handling on your side, because you would have been able to detect it by interpreting the error message.

- Another thing that I noticed is this screen after scanning QR or trying to connect via the deep link:


![Image](https://user-images.githubusercontent.com/23366251/225568707-0497cb35-1bb6-4ecc-b05a-8bacd206a8d6.jpg)

I was not able to get any logs from the BitKeep wallet, so please collect them so we can figure out what's the reason of having this screen from time to time.

**Comment by china-mobile2008** [comment]
hey @jakubuid 

 I captured some error messages of WalletConnect links to BitKeep, hoping they will be helpful for you.

[bitkeep.log](https://github.com/WalletConnect/WalletConnectKotlinV2/files/10999641/bitkeep.log)

**Comment by jakubuid** [comment]
Hey, I did some changes in the validation logic and if you were to build Kotlin Dapp from develop, there is no problem with establishing a session.

There is a problem with the connection though. Indeed it takes a lot of time to initialize the connection, especially in the deep link flow when I switch to BitKeep wallet.

Regarding that I have a couple of things:
- from the logs, I can see that you're not using the latest version, please update to the BOM 1.6.1 and let me know if it helps with the connection initialization issue
- secondly, could you please tell me when you initialize the SDK in regard to the splash screen animation? does the initialization start when the animation starts as well? thanks to this you could safe the initialization time and initialize the SDK underneath the animation in parallel

**Comment by china-mobile2008** [comment]
@jakubuid 
The 1.6.1 version of the official website cannot connect to the BitKeep 7.3.0 version wallet, so I used the Walletconnect Bom_1.4.1. Below is the log information for the 1.6.1 version of the official website.

![image](https://user-images.githubusercontent.com/10908640/226235490-c5647187-7204-415e-9460-2a9996bbf6a8.png)

[bom_1.6.1_bitkeep.log](https://github.com/WalletConnect/WalletConnectKotlinV2/files/11013225/bom_1.6.1_bitkeep.log)

**Comment by jakubuid** [comment]
Hi, I think I already mentioned the solution for the issue in this comment: https://github.com/WalletConnect/WalletConnectKotlinV2/issues/687#issuecomment-1471570774

Please apply those suggestions and let me know the results.

The logs that you sent are missing the most important information which is what you are sending in `Sign.Params.Approve`.

Again, apply my suggestion and provide what you are sending with`Sign.Params.Approve`.

**Comment by jakubuid** [comment]
@china-mobile2008 are you still seeing those issues or we can close this ticket?

**Comment by china-mobile2008** [comment]
@jakubuid 

I can connect to BitKeep using BOM version 1.4.1. However, since you modified the namespace structure in versions 1.4.1+, and the latest version 7.3.1 of BitKeep still uses the old namespace structure, I will wait for BitKeep to upgrade Walletconnect before continuing to verify this issue. Thank you very much for your support, and this issue can be closed now.

**Comment by jakubuid** [comment]
Got it. Let me know if anything! :)
